File: 1662.check-if-two-string-arrays-are-equivalent.cpp
/*
 * @lc app=leetcode id=1662 lang=cpp
 *
 * [1662] Check If Two String Arrays are Equivalent
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  bool arrayStringsAreEqual(vector<string>& word1, vector<string>& word2) {
    int total1 = accumulate(
        begin(word1), end(word1), 0,
        [](int sum, const auto& word) { return sum + word.length(); });
    int total2 = accumulate(
        begin(word2), end(word2), 0,
        [](int sum, const auto& word) { return sum + word.length(); });
    if (total1 != total2) {
      return false;
    }
    int widx1 = 0, widx2 = 0, cidx1 = 0, cidx2 = 0;
    int len2 = word2[widx2].length();
    for (widx1 = 0; widx1 < word1.size(); widx1++) {
      int len1 = word1[widx1].length();
      for (cidx1 = 0; cidx1 < len1; cidx1++) {
        if (word1[widx1][cidx1] != word2[widx2][cidx2]) {
          return false;
        }
        cidx2++;
        if (cidx2 >= len2) {
          widx2++;
          if (widx2 >= word2.size()) {
            return true;
          }
          cidx2 = 0;
          len2 = word2[widx2].length();
        }
      }
    }
    return true;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<string> word1, word2;
  word1 = {"ab", "c"};
  word2 = {"a", "bc"};
  ASSERT_EQ(s.arrayStringsAreEqual(word1, word2), true);
  word1 = {"a", "cb"};
  word2 = {"ab", "c"};
  ASSERT_EQ(s.arrayStringsAreEqual(word1, word2), false);
  word1 = {"abc", "d", "defg"};
  word2 = {"abcddefg"};
  ASSERT_EQ(s.arrayStringsAreEqual(word1, word2), true);
  return 0;
}

File: 1662.check-if-two-string-arrays-are-equivalent.rs
/*
 * @lc app=leetcode id=1662 lang=rust
 *
 * [1662] Check If Two String Arrays are Equivalent
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn array_strings_are_equal(word1: Vec<String>, word2: Vec<String>) -> bool {
        let mut word1_combined = String::new();
        for s in word1 {
            word1_combined += &s;
        }
        let mut word2_combined = String::new();
        for s in word2 {
            word2_combined += &s;
        }
        word1_combined == word2_combined
    }
}
// @lc code=end

#[test]
fn test1() {
    assert!(Solution::array_strings_are_equal(
        vec!["ab".to_string(), "c".to_string()],
        vec!["a".to_string(), "bc".to_string()]
    ));
}

#[test]
fn test2() {
    assert!(!Solution::array_strings_are_equal(
        vec!["a".to_string(), "cb".to_string()],
        vec!["ab".to_string(), "c".to_string()]
    ));
}

#[test]
fn test3() {
    assert!(Solution::array_strings_are_equal(
        vec!["abc".to_string(), "d".to_string(), "defg".to_string()],
        vec!["abcddefg".to_string()]
    ));
}
